rust/sdk/src/client.rs

```
@@ -388,6 +316,356 @@ pub(crate) fn to_lines(stream: impl Stream<Item = Result<Bytes, reqwest::Error>>
+    /// { "lamport": 1, "event": { "temperature": 10 } }
+    /// { "lamport": 3, "event": { "temperature": 12 } }
+    /// { "lamport": 14, "event": { "temperature": 9 } }
+    /// ```
+    /// If you set the lower bound to `10`, only the last event will be returned.
+    pub fn with_lower_bound(mut self, lower_bound: OffsetMap) -> Self {
+        if let Self::Initial { ref mut request, .. } = self {
+            request.lower_bound = Some(lower_bound);
+        }
+        self
+    }
+
+    /// Add an upper bound to the query.
+    ///
+    /// The upper bound limits the start of the query events.
+    /// As an example, consider the following (example) events:
+    /// ```json
+    /// { "lamport": 1, "event": { "temperature": 10 } }
+    /// { "lamport": 3, "event": { "temperature": 12 } }
+    /// { "lamport": 14, "event": { "temperature": 9 } }
+    /// ```
+    /// If you set the upper bound to `10`, the first two events will be returned.
+    pub fn with_upper_bound(mut self, upper_bound: OffsetMap) -> Self {
+        if let Self::Initial { ref mut request, .. } = self {
+            request.upper_bound = Some(upper_bound);
+        }
+        self
+    }
+
+    /// Set the query's event order. By default, this value is set to [`Order::Asc`].
+    ///
+    /// As an example, consider the following (example) events:
+    /// ```json
+    /// { "lamport": 1, "event": { "temperature": 10 } }
+    /// { "lamport": 3, "event": { "temperature": 12 } }
+    /// { "lamport": 14, "event": { "temperature": 9 } }
+    /// ```
+    /// If your query sets [`Order::Desc`], the result will instead look like:
+    /// ```json
+    /// { "lamport": 14, "event": { "temperature": 9 } }
+    /// { "lamport": 3, "event": { "temperature": 12 } }
+    /// { "lamport": 1, "event": { "temperature": 10 } }
+    /// ```
+    pub fn with_order(mut self, order: Order) -> Self {
+        if let Self::Initial { ref mut request, .. } = self {
+            request.order = order;
+        }
```

I’m on the fence, but slightly leaning towards panicking in the `else` branch.

---

How could it happen? 🤔 Could someone be messing with the internals?
Like on the `Empty` case from the future, I prefer crashing than silently failing

---

Well, someone would have to

- create the Future
- manually poll it once
- then call such a method

This is so clearly an error that a panic seems justified. These Futures should be consumed using `.await`, in which case no such problem can occur.